src/index.ts

```diff
@@ -1,35 +1,41 @@
+import R from 'ramda'
+
 import { load } from './load'
 import { validate } from './validate'
+import { validate as validateParameters } from './params/validate'
 
 import { MemoizedConfig } from './types'
-
-interface Parameters {
-  // TODO: define Parameters interface
-}
+import { defaultParameters, Parameters } from './params'
 
 export = class StrongConfig {
-  public readonly params: Parameters | undefined
+  public readonly parameters: Parameters
   private config: MemoizedConfig
 
-  constructor(params?: Parameters) {
-    // TODO: validate params
-    this.params = params
+  constructor(parameters?: Partial<Parameters>) {
+    this.parameters = validateParameters(
```

Comment:
Thanks for the links. I still think the constructor is the best place.

The links seem to more refer to _side effects_, which is not what we have. I like the current contract of _if you pass valid params, you can create a strong-config instance. If not, you can't_. That's deterministic and is not a side-effect.

Additionally, validating in the methods mean we duplicate the validation logic in multiple places instead of validation once for one instance (which is sufficient since params are not designed to change at runtime).

Reply:
Thanks for the detailed reply.

I agree with
- constructors shouldn't do much
- no async action possible in constructors, single threaded JS, etc.
- no side effects (no disk IO, no http calls, etc.)

but disagree with all the rest though
- we're not really reading a file in the constructor a la `fs.readFile`. I mean it's just another import: `import { parametersSchema } from './schema'`, see `params/validate.ts`.
- Why much harder to test in our case? I test the `validateParameters` isolatedly, and, assert on `strongConfig.parameters` when constructing the class. Straight forward unit tests.
- destructors assume a previously constructed object. If construction fails, there can't be a leak. Lacking in-depth knowledge here but the links below support my guess.
- Nothing is 'bloated' in this constructor. User input validation is not an anti-pattern for me, including constructor logic. 

In fact, throwing inside a constructor seems to be totally fine in programming

- https://stackoverflow.com/questions/6086334/is-it-good-practice-to-make-the-constructor-throw-an-exception/6086399
- https://stackoverflow.com/questions/810839/throwing-exceptions-from-constructors
- https://stackoverflow.com/questions/5149619/what-should-a-javascript-constructor-return-if-it-fails
- https://stackoverflow.com/questions/32323406/what-happens-if-a-constructor-throws-an-exception/32323458

Reply:
After doing a lot more reading on the topic it definitely seems like there's two schools of thought with regards to the right way to do this. Pros and cons to both it looks like.

In this particular case, I would say that you're right. But to address your points:

- I believe that JS imports are file IO...but I can't find anything good here to back up my claim
- Testing difficulty comes when we introduce [subclassing](https://www.oreilly.com/library/view/learning-javascript-design/9781449334840/ch09s12.html)

An example:

```
class A {
  constructor() {
    someFuncThatErrors()
  }
}

class B extends A {
  constructor() {
  }
}
```

When testing A you can simply mock behaviour from `someFuncThatErrors` and successfully assert. When testing B you can't so easily (though I believe it's not too bad in JS). 

- For your point with regards to destructors, memory is allocated *before* object instantiation. If destructors aren't called, this memory is sometimes not cleaned up. I'm also lacking some deep JS knowledge here to know if this is the case, but this is a key point in other languages.
- You're right about user input validation, but I'm sure you'll agree how user input validation can go too far:

```
class User {
  constructor(userId) {
     checkUserExistsInDB(userId)
  }
} 
```

Technically, user input validation (we don't want to construct objects with invalid user IDs). 

Basically what I'm saying is, that for now, your approach is fine, but this pattern can very quickly turn messy and sometimes it's better to just allow objects to instantiate and defer work until later.

Happy to resolve for now, once we've concluded the discussion :)

Reply:
Thanks again for the details, much appreciated! And yes, I agree.

If we just had to choose between `validating in the constructor` and `validate in the method` without anything else to consider, i'd go for the method, fully agree. My major point in our case though is DRY. Think it's much nicer to validate once where you only need to validate once.